tests/OutputTest/OutputTest.php

```diff
@@ -0,0 +1,209 @@
+        }
+    }
+
+    /**
+     * @dataProvider outputTestProvider
+     */
+    public function testOutputMatchesReferenceRendering(Dataset $dataset): void
+    {
+        $document = $dataset->render();
+        $referenceFile = $dataset->referenceFile()->getPathname();
+        $actualOutputFile = tempnam(sys_get_temp_dir(), "dompdf_test_");
+
+        file_put_contents($actualOutputFile, $document->output());
+
+        try {
+            $this->assertOutputMatches($referenceFile, $actualOutputFile);
+        } catch (AssertionFailedError $e) {
+            $this->saveFailedOutput($dataset);
+            throw $e;
+        } finally {
+            unlink($actualOutputFile);
+        }
+    }
+
+    private function assertOutputMatches(
+        string $referenceFile,
+        string $actualOutputFile
+    ): void {
+        $command = function ($file) {
+            return [
+                "gs",
+                "-q", "-dBATCH", "-dNOPAUSE", "-sstdout=%stderr",
+                "-sDEVICE=png16m", "-dGraphicsAlphaBits=4",
+                "-sOutputFile=-", $file
+            ];
+        };
+        $process1 = new Process($command($referenceFile));
+        $process2 = new Process($command($actualOutputFile));
+
+        $process1->mustRun();
+        $process2->mustRun();
+
+        $error1 = $process1->getErrorOutput();
+        $error2 = $process2->getErrorOutput();
+
+        if ($error1 !== "") {
+            throw new RuntimeException("Ghostscript execution failed: $error1");
+        }
+
+        if ($error2 !== "") {
+            throw new RuntimeException("Ghostscript execution failed: $error2");
+        }
+
+        $referenceImages = $this->outputToImageData($process1->getOutput());
+        $actualImages = $this->outputToImageData($process2->getOutput());
+
+        $expectedCount = count($referenceImages);
+        $failureMessage = "Output does not match reference rendering (number of pages is different).";
+        $this->assertCount($expectedCount, $actualImages, $failureMessage);
+
+        foreach ($referenceImages as $page => $referenceData) {
+            $actualData = $actualImages[$page];
+
+            $matches = $this->compareImages($referenceData, $actualData);
+            $failureMessage = "Output does not match reference rendering (difference on page $page).";
+            $this->assertTrue($matches, $failureMessage);
+        }
+    }
+
+    /**
+     * @param string $output The Ghostscript command output, containing PNG image data.
+     *
+     * @return string[] A list of the PNG images defined by the output.
+     */
+    private function outputToImageData(string $output): array
+    {
+        $pngSignature = "\x89\x50\x4E\x47\x0D\x0A\x1A\x0A";
+        $elements = explode($pngSignature, $output);
+
+        if (count($elements) <= 1 || $elements[0] !== "") {
+            throw new RuntimeException("Unexpected Ghostscript output: $output");
```

Comment:
```suggestion
            throw new RuntimeException("Unexpected Ghostscript output: `$output`");
```

I think its always nice to put quote around it then you know what the about was because you could get something like this: `Uncaught PHP Exception RuntimeException : "Unexpected Ghostscript output: " at /...` then you dont know if the output was ` ` or it was ` `

Reply:
Done.